File: core/ui/src/types/complaint.ts
import type { PageResponse } from "./grn";

export type ComplaintSource = "CUSTOMER" | "MARKET" | "CLINICAL" | "INTERNAL" | "DISTRIBUTOR" | "REGULATORY_AUTHORITY";
export type ComplaintCategory = "PRODUCT_QUALITY" | "ADVERSE_EVENT" | "LABELING_ERROR" | "PACKAGING_DEFECT" | "EFFICACY" | "CONTAMINATION" | "OTHER";
export type ComplaintSeverity = "CRITICAL" | "MAJOR" | "MINOR" | "INFORMATIONAL";
export type ComplaintStatus = "RECEIVED" | "UNDER_INVESTIGATION" | "PENDING_CLOSURE" | "CLOSED" | "WITHDRAWN";
export type RegulatoryReportability = "NOT_ASSESSED" | "REPORTABLE" | "NOT_REPORTABLE" | "REPORTED";

export type Complaint = {
  id: string;
  complaintNumber: string;
  receivedDate: string;
  source: ComplaintSource;
  category: ComplaintCategory;
  severity: ComplaintSeverity;
  status: ComplaintStatus;
  productName: string | null;
  lotNumber: string | null;
  reportedBy: string | null;
  description: string;
  initialAssessment: string | null;
  rootCause: string | null;
  impactAssessment: string | null;
  recallRequired: boolean;
  regulatoryReportability: RegulatoryReportability;
  regulatoryReportDate: string | null;
  regulatoryAuthority: string | null;
  linkedDeviationId: string | null;
  linkedCapaId: string | null;
  closedBy: string | null;
  closedAt: string | null;
  closureSummary: string | null;
  isActive: boolean;
  createdBy: string;
  createdAt: string;
  updatedBy: string | null;
  updatedAt: string | null;
};

export type CreateComplaintRequest = {
  description: string;
  receivedDate: string;
  source: ComplaintSource;
  category: ComplaintCategory;
  severity: ComplaintSeverity;
  productName?: string;
  lotNumber?: string;
  reportedBy?: string;
  initialAssessment?: string;
};

export type UpdateComplaintRequest = {
  category: ComplaintCategory;
  severity: ComplaintSeverity;
  description: string;
  productName?: string;
  lotNumber?: string;
  reportedBy?: string;
  initialAssessment?: string;
  rootCause?: string;
  impactAssessment?: string;
  recallRequired: boolean;
  regulatoryReportability: RegulatoryReportability;
  regulatoryAuthority?: string;
  regulatoryReportDate?: string;
};

export type ComplaintStatusUpdateRequest = {
  status: ComplaintStatus;
  reason?: string;
  closureSummary?: string;
  username?: string;
  password?: string;
  meaning?: string;
};

export type ComplaintSummary = {
  countsByStatus: Record<ComplaintStatus, number>;
  countsByCategory: Record<ComplaintCategory, number>;
  countsBySeverity: Record<ComplaintSeverity, number>;
};

export type ComplaintPage = PageResponse<Complaint>;

File: core/src/main/java/com/batchsphere/core/qms/complaint/entity/Complaint.java
package com.batchsphere.core.qms.complaint.entity;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.UUID;

@Entity
@Table(name = "complaint")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class Complaint {

    @Id
    @Column(nullable = false, updatable = false)
    private UUID id;

    @Column(name = "complaint_number", nullable = false, unique = true, length = 30)
    private String complaintNumber;

    @Column(name = "received_date", nullable = false)
    private LocalDate receivedDate;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ComplaintSource source;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ComplaintCategory category;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    private ComplaintSeverity severity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ComplaintStatus status;

    @Column(name = "product_name", length = 255)
    private String productName;

    @Column(name = "lot_number", length = 100)
    private String lotNumber;

    @Column(name = "reported_by", length = 255)
    private String reportedBy;

    @Column(nullable = false, columnDefinition = "TEXT")
    private String description;

    @Column(name = "initial_assessment", columnDefinition = "TEXT")
    private String initialAssessment;

    @Column(name = "root_cause", columnDefinition = "TEXT")
    private String rootCause;

    @Column(name = "impact_assessment", columnDefinition = "TEXT")
    private String impactAssessment;

    @Column(name = "recall_required")
    private Boolean recallRequired;

    @Enumerated(EnumType.STRING)
    @Column(name = "regulatory_reportability", length = 20)
    private RegulatoryReportability regulatoryReportability;

    @Column(name = "regulatory_report_date")
    private LocalDate regulatoryReportDate;

    @Column(name = "regulatory_authority", length = 100)
    private String regulatoryAuthority;

    @Column(name = "linked_deviation_id")
    private UUID linkedDeviationId;

    @Column(name = "linked_capa_id")
    private UUID linkedCapaId;

    @Column(name = "closed_by", length = 100)
    private String closedBy;

    @Column(name = "closed_at")
    private OffsetDateTime closedAt;

    @Column(name = "closure_summary", columnDefinition = "TEXT")
    private String closureSummary;

    @Column(name = "is_active", nullable = false)
    private Boolean isActive;

    @Column(name = "created_by", nullable = false, length = 100)
    private String createdBy;

    @Column(name = "created_at", nullable = false)
    private OffsetDateTime createdAt;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Column(name = "updated_at")
    private OffsetDateTime updatedAt;
}
